internal/connection/singleton.go

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package connection
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"sync"
+	"time"
+
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/hazelcast/hazelcast-go-client"
+	"github.com/hazelcast/hazelcast-go-client/sql/driver"
+
+	hzcerrors "github.com/hazelcast/hazelcast-commandline-client/errors"
+)
+
+// being initialized at compile-time.
+var (
+	GitCommit     string
+	ClientVersion string
+)
+
+const clientResponseTimeoutDeadline = 1 * time.Second
+
+type singletonHZClient struct {
+	mu        sync.Mutex
+	sqlDriver *sql.DB
+	client    *hazelcast.Client
+}
+
+var clientLock = &sync.Mutex{}
+var hzInstance *singletonHZClient
+
+func newSingletonHZClient(ctx context.Context, clientConfig *hazelcast.Config) (*singletonHZClient, error) {
+	var err error
+	sc := &singletonHZClient{}
+	configCopy := clientConfig.Clone()
+	sc.client, err = hazelcast.StartNewClientWithConfig(ctx, configCopy)
+	if err != nil {
+		return nil, err
+	}
+	return sc, nil
+}
+
+func getHZClientInstance(ctx context.Context, clientConfig *hazelcast.Config) (*singletonHZClient, error) {
+	var err error
+	if hzInstance == nil {
```

Comment:
`hzInstance` should be protected for double checked lock to work, but why is the double checked lock necessary in the first place?

Reply:
I think the pre-located locking is not required for double lock.
Ref: https://refactoring.guru/design-patterns/singleton/go/example#example-0--single-go.